reader_concurrency_semaphore.cc

```diff
@@ -1256,9 +1256,11 @@ std::string reader_concurrency_semaphore::dump_diagnostics(unsigned max_lines) c
     return os.str();
 }
 
-void reader_concurrency_semaphore::foreach_permit(noncopyable_function<void(const reader_permit&)> func) {
+void reader_concurrency_semaphore::foreach_permit(noncopyable_function<void(const reader_permit&)> func) const {
     for (auto& p : _permit_list) {
-        func(reader_permit(p.shared_from_this()));
+        // We cast away const to construct a reader_permit but the resulting
+        // object is passed as const& so there is no const violation here.
+        func(reader_permit(const_cast<reader_permit::impl&>(p).shared_from_this()));
```

Comment:
`const reader_permit&` is a lie. Suppose you have one, you can easily create a mutable reader_permit referring to the same underlying permit using the copy constructor, and thus bypass const protection without const_cast. That's why we have two types for shared_ptr<T> and shared_ptr<const T>, and you can't construct the mutable one from a const one.

I don't think it's worth fixing it though. Does constness protect anything for us?

Reply:
> const reader_permit& is a lie. 
> I don't think it's worth fixing it though.

I agree.

> Does constness protect anything for us?

Protect is a strong word for what a const reader_permit& does. It doesn't allow consuming resources. But it is very easy to circumvent like you said. Still, a well behaved developer should not circumvent it.

Reply:
The point is it can get circumvented unknowingly, just by copying.